tests/integration/testlib/iproutelib.py

```diff
@@ -0,0 +1,81 @@
+#
+# Copyright 2019 Red Hat, Inc.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+
+from contextlib import contextmanager
+import subprocess
+import threading
+import time
+
+
+TIMEOUT = 10
+
+
+class IpMonitorResult(object):
+    def __init__(self):
+        self.out = None
+        self.err = None
+        self.popen = None
+
+
+@contextmanager
+def ip_monitor(object_type, dev, timeout=10):
+    result = IpMonitorResult()
+
+    cmds = 'timeout {} ip monitor {} dev {}'.format(timeout, object_type, dev)
+
+    def run():
+        result.popen = subprocess.Popen(
+            cmds.split(),
+            close_fds=True,
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            env=None
+        )
+        result.out, result.err = result.popen.communicate(None)
+
+    def finalize():
+        if result.popen:
+            result.popen.terminate()
+
+    with _thread(run, 'ip-monitor', teardown_cb=finalize):
+        # Let the ip monitor thread start before proceeding to the action.
+        time.sleep(1)
+        yield result
+
+
+def get_non_up_events(result, dev):
+    """
+    Given a result and device, filter only the non UP events (DOWN, UNKNOWN)
+    and return them as a list.
+    :param result: IpMonitorResult
+    :return: List of non UP events
+    """
+    return [l for l in result.out.split('\n')
+            if 'state UP' not in l and dev in l]
```

Comment:
You might need to check NO-CARRIER for software interface.

When adding interface to bond, I noticed this state:
```
bond99: <NO-CARRIER,BROADCAST,MULTICAST,MASTER,UP>
```

Reply:
event example:
`eth1: <BROADCAST,MULTICAST> mtu 1500 qdisc pfifo_fast state DOWN group default`

We are looking for the `state DOWN` in this case.
The `NO-CARRIER` and the other  properties in the `<..>` are about the current flags of the interface. I think one is the current state and the other is the transition event.
Have you seen this affecting the `state ????` value?

Reply:
The reapply does not allow changes on `connection.master`, so it seems bond cannot benefit on reapply:
```
[fge@fed nmstate]$ sudo nmcli d reapply eth2
Error: Reapplying connection to device 'eth2' (/org/freedesktop/NetworkManager/Devices/5) failed: Can't reapply changes to 'connection.master' setting
```

Reply:
Rebinding a slave to a bond or bridge is disruptive, but changing an existing bond or adding a new slave should probably be supported (while checking the bond/bridge itself).